### PR TITLE
Teleport with play area

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
- Teleport beam.
  - The teleport parabola is visible behind entities.
  - The teleport target overlay has changed.
  - The VR play area is (optionally) displayed about the teleport destination.
  - Parts of the teleport target and play area obscured by entities are visible.
- Settings > Controllers... > VR Movement options:
  - "Advanced movement" and "jumping and flying" options reworked.
  - "Jumping and flying" only available if "walking and teleporting" selected.
  - "Snap turn" and "smooth turn" options layout changed.
  - "Show room boundaries" option added that enables/disables the display of the VR play area. Includes mention of sensors if Oculus HMD.
  - "Reset Sensors" button removed.
- API changes:
  - MyAvatar.showPlayArea property added.
  - drawInFront property added to parabola pointer.
  - HMD.playArea property added that reports the size and position of the play area rectangle.
  - HMD.sensorsPositions property that reports the positions of the Oculus sensors.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13266/HighFidelity-Beta-PR14075-90ef109.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13266/HighFidelity-Beta-PR14075-90ef109.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13266/HighFidelity-Beta-Interface-PR14075-90ef109.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13266/HighFidelity-Beta-PR14075-90ef109.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13266/HighFidelity-Beta-Interface-PR14075-90ef109.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13444/HighFidelity-Beta-PR14075-545c259.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13444/HighFidelity-Beta-PR14075-545c259.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13444/HighFidelity-Beta-Interface-PR14075-545c259.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13446/HighFidelity-Beta-PR14075-1c4bde3.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13446/HighFidelity-Beta-PR14075-1c4bde3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13446/HighFidelity-Beta-Interface-PR14075-1c4bde3.dmg

**Comment by ctrlaltdavid** [comment]
### Test Plan

A) Oculus HMD

- The teleport pad has changed from the word "teleport" to an image of feet.
- When you activate your teleport beam, you can see a square representing your play area on the ground surrounding your teleport pad: if and only if Settings > Controllers... > Show Room Boundaries is enabled.
- Check that the play area corresponds to the Oculus play area: correct size and location. (E.g., walk around the edge of the play area so that the Oculus software displays the boundaries and check that the play area displayed matches.)
- Your sensors are also visible at their accurate position and orientation (Note: they are shown in their position and orientation _at the time of setup_; if they've moved since their positions won't update automatically.)
- Check the play area and sensors are correctly sized and positioned for different avatar scales.
- When your teleport pad or play area are occluded by geometry, the occluded portions are visible in a darker blue.
- The teleport beam is visible through geometry.
- If you have not set up a play area in the Oculus software it will not show.
- Test the Settings > Controllers... > "Walking and teleporting" and "Walking only" options. (These replace the prior "Advanced movement" option.
- Test that you can only jump and fly if "Walking and teleporting" plus "Jumping and flying" are selected.
- The teleport pad, play area, and sensors continue to display for a short while about your new location after you have teleported. These are immediately hidden if you turn on the teleport beam again with either hand, or you start walking.

B) Vive HMD: Repeat the A tests, except that sensor positions are not displayed.

C) Microsoft MR HMD: Repeat the A tests, , except that sensor positions are not displayed.

D) Ad-hoc testing
- Use this PR in several domains, teleport to a variety of heights and on rough terrain.
- Use this PR with play areas of various sizes.
- Check the PR features after entering, exiting, and re-entering VR mode.
- Reload all scripts while in HMD mode. Check that the teleporting still works.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13484/HighFidelity-Beta-PR14075-7560a2a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13484/HighFidelity-Beta-PR14075-7560a2a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13484/HighFidelity-Beta-Interface-PR14075-7560a2a.dmg

**Comment by qlimguicheteau** [comment]
Now Testing

**Comment by qlimguicheteau** [comment]
Testing Paused
Waiting on Microsoft Mixed reality results.

**Comment by qliejahner** [comment]
Fails on Rift, Microsoft MR results pending

Play area does not scale correctly to small avatars. The play area remains approximately the same size relative to the world, but as a smaller avatar, the user cannot travel that distance before reaching the edge of their play area. When switching from the mannequin to Robimo, the projected play area appears much larger than before.

If an edge of the play area represents a wall in real life, it can be lined up to a wall in high fidelity, and this discrepancy becomes more obvious.

This has to do with the size of the avatar relative to the mannequin; the projection remains the same when increasing or decreasing scale.

hifi-snap-by--on-2018-10-01_11-49-12
hifi-snap-by--on-2018-10-01_11-49-45


hifi-snap-by--on-2018-10-01_11-43-29
hifi-snap-by--on-2018-10-01_11-45-54


With smaller avatars, the diagram is always partially occluded on flat surfaces, meaning the borders are not visible after teleporting. This occurs on avatars which are only slightly shorter than default, such as the anime boy.

hifi-snap-by--on-2018-10-01_11-49-56

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13533/HighFidelity-Beta-PR14075-87ff6dd.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13533/HighFidelity-Beta-PR14075-87ff6dd.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13533/HighFidelity-Beta-Interface-PR14075-87ff6dd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13533/HighFidelity-Beta-PR14075-87ff6dd.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13533/HighFidelity-Beta-Interface-PR14075-87ff6dd.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13549/HighFidelity-Beta-PR14075-84a06fb.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13549/HighFidelity-Beta-PR14075-84a06fb.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13549/HighFidelity-Beta-Interface-PR14075-84a06fb.dmg

**Comment by ctrlaltdavid** [comment]
PR update with fixes:
- Fix gaps between visible and occluded parts of play area overlays.
- Fix occluded parts of play area overlays not displaying for a time after teleport.

Ready for QA.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13552/HighFidelity-Beta-PR14075-f7fd76b.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13552/HighFidelity-Beta-PR14075-f7fd76b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13552/HighFidelity-Beta-Interface-PR14075-f7fd76b.exe

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
Pass? @conklin94122 
The test plan passes as written.
However, there are a few small flaws which might constitute a failure or bugs. @ctrlaltdavid 

The non-occluded play area border is not visible from below.
hifi-snap-by--on-2018-10-02_09-59-56

The transparent ring around the teleport target hides the occluded blue parts of the teleport target.
hifi-snap-by--on-2018-10-02_09-52-06

As expected, the diagram is hidden if the user moves, flies, or teleports, which would cause it to be inaccurate. However, it is not hidden if there is significant movement from an outside force (such as slopes or moving floors) or if the user rotates with the controller.
hifi-snap-by--on-2018-10-02_09-39-03 1

File: scripts/system/controllers/controllerModules/teleport.js [hunk 1]
```diff
@@ -150,19 +149,149 @@ Script.include("/~/system/libraries/controllers.js");
         this.teleportParabolaHeadVisuals;
         this.teleportParabolaHeadCollisions;
 
+
+        this.PLAY_AREA_OVERLAY_MODEL = Script.resolvePath("../../assets/models/trackingSpacev9.fbx");
```
Comment (danteruiz) [inline scripts/system/controllers/controllerModules/teleport.js]: why aren't these constants at the top of the file like the others you added?

**Comment by ctrlaltdavid** [comment]
@JeffClinton Could you please check the colours of the play area and sensor overlays in the updated PR?  They should be semi-transparent cyan, not black. (The occluded parts should still be the same blue, though.)

File: scripts/system/controllers/controllerModules/teleport.js (hunk 1 above)
Comment (ctrlaltdavid) [reply scripts/system/controllers/controllerModules/teleport.js]: I have intentionally grouped these play area constants together with related variables and code.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13583/HighFidelity-Beta-PR14075-2bdd8bf.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13583/HighFidelity-Beta-PR14075-2bdd8bf.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13583/HighFidelity-Beta-Interface-PR14075-2bdd8bf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13583/HighFidelity-Beta-PR14075-2bdd8bf.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13583/HighFidelity-Beta-Interface-PR14075-2bdd8bf.exe

**Comment by JeffClinton** [comment]
QA Approved tag removed since changes were made.

@ctrlaltdavid If this is ready for testing again can you add the QA Ready tag?  That way the first available tester can grab it.  Thank you! :)

**Comment by ctrlaltdavid** [comment]
Evan's going to do the two tests tomorrow.

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
Passing here; no such stripes appear on Rift. Border is no longer black, and can be seen from below.

**Pull request merged**